### PR TITLE
Expose the U3 operation to frontends

### PR DESCRIPTION
**Pull request by khalatepradnya**
### Description
Partially addresses issue https://github.com/NVIDIA/cuda-quantum/issues/1507

### CODE REVIEW

File: lib/Frontend/nvqpp/ConvertExpr.cpp
Comment (schweitzpgi) [inline lib/Frontend/nvqpp/ConvertExpr.cpp:1597]: Can you add a comment here? Something like `&& "The u3 operator must have 4 arguments"` would be great.

File: python/cudaq/kernel/ast_bridge.py
Comment (schweitzpgi) [inline python/cudaq/kernel/ast_bridge.py:1372]: Not a big deal, but we expect exactly 3 angles here so it would be straightforward to create the params without the `reverse()`.

```python
lambd4 = self.popValue()
phi = self.popValue()
theta = self.popValue()
params = [theta, phi, lambd4]
```

I assume there is no way to pop a vector of values in natural order like in the C++ bridge.

```c++
std::vector<Value> args = popValues(3);
```

File: python/tests/builder/test_kernel_builder.py
Comment (schweitzpgi) [inline python/tests/builder/test_kernel_builder.py:1250]: Does this mean you want python to accept a list of qubits in the final position? I assume that would be checked such that the list must contain exactly one qubit or it is ill-formed.

File: runtime/cudaq/qis/qubit_qis.h
Comment (schweitzpgi) [inline runtime/cudaq/qis/qubit_qis.h:39]: Should we be adding `u2` here while we're at it?

File: runtime/nvqir/Gates.h
Comment (schweitzpgi) [inline runtime/nvqir/Gates.h:117]: @boschmitt

File: runtime/nvqir/NVQIR.cpp
Comment (schweitzpgi) [inline runtime/nvqir/NVQIR.cpp:335]: We should find a way of specifying this formula in a single place. It keeps repeating over and over.

**Review by schweitzpgi — COMMENTED** [review]
Looks great. Just a few questions.

File: runtime/nvqir/NVQIR.cpp
Comment (1tnguyen) [reply runtime/nvqir/NVQIR.cpp:335]: In `runtime/nvqir/Gates.h`, we have the `getGateByName` support for U3.

File: lib/Frontend/nvqpp/ConvertExpr.cpp
Comment (amccaskey) [reply lib/Frontend/nvqpp/ConvertExpr.cpp:1597]: Also should we not be leveraging `isControl` and/or `negations` and/or `isAdjoint`?

File: python/tests/kernel/test_kernel_features.py
Comment (amccaskey) [inline python/tests/kernel/test_kernel_features.py:1286]: would be good to add a test that takes the 3 angles as input to the kernel

File: runtime/cudaq/builder/kernel_builder.cpp
Comment (amccaskey) [inline runtime/cudaq/builder/kernel_builder.cpp:569]: We'll need the control version too.

Comment (schweitzpgi) [reply runtime/cudaq/builder/kernel_builder.cpp:569]: Do we want/need a `cu3` for C++17?

File: lib/Optimizer/Dialect/Quake/QuakeOps.cpp
Comment (boschmitt) [inline lib/Optimizer/Dialect/Quake/QuakeOps.cpp:899]: This is indeed a column-major representation, while the one in `runtime/nvqir/NVQIR.cpp` is a row-major one. Matrices must be in row-major! (Hence this one, is wrong.)

File: runtime/nvqir/Gates.h
Comment (boschmitt) [reply runtime/nvqir/Gates.h:117]: It's wrong, this matrix must be in row-major representation.

File: runtime/nvqir/NVQIR.cpp
Comment (boschmitt) [reply runtime/nvqir/NVQIR.cpp:335]: +1

**Review by boschmitt — COMMENTED** [review]
For historical reasons the expected matrix representation changed from column-major to row-major. (It has to do with different backends expecting different forms by default.) In any case, for most operators, this changes nothing but in the case of `u3` the representation is relevant.

As we see in this PR, we have `u3` matrix definition in various places and with different representations! We need to find a way to centralize operators matrix definitions and make sure they are all in row-major form.

File: runtime/cudaq/qis/qubit_qis.h
Comment (khalatepradnya) [reply runtime/cudaq/qis/qubit_qis.h:39]: No, I don't plan to. U2 can be implemented with U3 if required.

File: lib/Optimizer/Dialect/Quake/QuakeOps.cpp
Comment (khalatepradnya) [reply lib/Optimizer/Dialect/Quake/QuakeOps.cpp:899]: Looking at other operations in this file, it seems to me that they all are column-major, for e.g. https://github.com/NVIDIA/cuda-quantum/blob/main/lib/Optimizer/Dialect/Quake/QuakeOps.cpp#L909

File: python/tests/builder/test_kernel_builder.py
Comment (khalatepradnya) [reply python/tests/builder/test_kernel_builder.py]: I changed the comment - the current behavior requires `q[0]`, asking if we should / need to support just using `q` and if `q` has multiple qubits, the operation is applied to all

File: runtime/nvqir/NVQIR.cpp
Comment (khalatepradnya) [reply runtime/nvqir/NVQIR.cpp]: Will address separately...

File: python/tests/builder/test_kernel_builder.py
Comment (khalatepradnya) [reply python/tests/builder/test_kernel_builder.py]: Fixed

File: python/tests/kernel/test_kernel_features.py
Comment (khalatepradnya) [inline python/tests/kernel/test_kernel_features.py:1380]: @boschmitt : Can you please check the Rx representation?

File: python/cudaq/kernel/ast_bridge.py
Comment (schweitzpgi) [inline python/cudaq/kernel/ast_bridge.py:1369]: Aside: at some point we should decide as a group how to write these angles. Code has gone from greek letters, to a mix of greek letters and English spellings of greek letters, to fully spelling them out in English (e.g. "angle_one"). I don't have a preference, but I don't think we need churn for the sake of churn.

**Review by schweitzpgi — APPROVED** [review]
Approving, contingent on Bruno double-checking that the matrices are the correct encodings.

File: python/tests/kernel/test_kernel_features.py
Comment (khalatepradnya) [reply python/tests/kernel/test_kernel_features.py]: Fixed in [20c9b87](https://github.com/NVIDIA/cuda-quantum/pull/1504/commits/20c9b87086e68f18af7438356ad22a71f70502b7)

File: python/cudaq/kernel/ast_bridge.py
Comment (khalatepradnya) [reply python/cudaq/kernel/ast_bridge.py:1369]: Tried to address in [d013595](https://github.com/NVIDIA/cuda-quantum/pull/1504/commits/d013595eacb390d5e6eb6b8044a1c7a1c325d7c3)
Did not change the comments within the code.

**Pull request merged**
